File: tts-pronunciations/python/example_pronunciation_dictionaries.py
#!/usr/bin/env python3
"""Manage a named pronunciation dictionary through Inworld's public REST API."""

import json
import os
import sys
from pathlib import Path
from typing import Any
from urllib.parse import quote
from uuid import UUID, uuid4

import requests

DEFAULT_API_BASE_URL = "https://api.inworld.ai"
REQUEST_TIMEOUT_SECONDS = 30

INITIAL_PRONUNCIATIONS = [
    {
        "displayHeadword": "Cat",
        "languageCode": "en-US",
        "phoneSymbols": ["k", "æ", "t"],
    },
    {
        "displayHeadword": "Ship",
        "languageCode": "en-US",
        "phoneSymbols": ["ʃ", "ɪ", "p"],
    },
]

REPLACEMENT_PRONUNCIATIONS = [
    {
        "displayHeadword": "Red",
        "languageCode": "en-US",
        "phoneSymbols": ["ɹ", "ɛ", "d"],
    }
]


class ApiError(RuntimeError):
    """An HTTP error with a status usable for bounded cleanup."""

    def __init__(self, method: str, status_code: int):
        self.status_code = status_code
        super().__init__(f"{method} request failed with HTTP {status_code}.")


class PronunciationDictionariesClient:
    """Minimal client for the five public pronunciation dictionary methods."""

    def __init__(self, api_key: str, workspace_id: str, api_base_url: str):
        self.api_base_url = api_base_url.rstrip("/")
        self.workspace_id = workspace_id
        encoded_workspace_id = quote(workspace_id, safe="")
        self.collection_url = (
            f"{self.api_base_url}/pronunciations/v1/workspaces/"
            f"{encoded_workspace_id}/pronunciationDictionaries"
        )
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Basic {api_key}",
                "Content-Type": "application/json",
            }
        )

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.session.close()

    def create_dictionary(
        self, display_name: str, pronunciations: list[dict[str, Any]]
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            self.collection_url,
            json={
                "displayName": display_name,
                "pronunciations": pronunciations,
            },
        )

    def list_dictionaries(self, page_size: int = 5) -> dict[str, Any]:
        return self._request(
            "GET",
            self.collection_url,
            params={"pageSize": page_size},
        )

    def get_dictionary(self, name: str) -> dict[str, Any]:
        return self._request("GET", self._resource_url(name))

    def update_dictionary(
        self,
        dictionary: dict[str, Any],
        display_name: str,
        pronunciations: list[dict[str, Any]],
    ) -> dict[str, Any]:
        require_etag(dictionary)
        return self._request(
            "PATCH",
            self._resource_url(dictionary["name"]),
            params={"updateMask": "displayName,pronunciations"},
            json={
                "displayName": display_name,
                "pronunciations": pronunciations,
                "etag": dictionary["etag"],
            },
        )

    def delete_dictionary(self, dictionary: dict[str, Any]) -> dict[str, Any]:
        require_etag(dictionary)
        return self._request(
            "DELETE",
            self._resource_url(dictionary["name"]),
            params={"etag": dictionary["etag"]},
        )

    def _resource_url(self, name: str) -> str:
        expected_prefix = f"workspaces/{self.workspace_id}/pronunciationDictionaries/"
        if not isinstance(name, str) or not name.startswith(expected_prefix):
            raise ValueError("Dictionary name must belong to INWORLD_WORKSPACE_ID.")
        try:
            UUID(name[len(expected_prefix) :])
        except ValueError:
            raise ValueError(
                "Dictionary name must end in the returned dictionary UUID."
            ) from None
        encoded_name = "/".join(quote(segment, safe="") for segment in name.split("/"))
        return f"{self.api_base_url}/pronunciations/v1/{encoded_name}"

    def _request(self, method: str, url: str, **kwargs: Any) -> dict[str, Any]:
        try:
            response = self.session.request(
                method,
                url,
                timeout=REQUEST_TIMEOUT_SECONDS,
                allow_redirects=False,
                **kwargs,
            )
        except requests.RequestException as error:
            raise RuntimeError(
                f"{method} request failed ({type(error).__name__}); its outcome may be unknown."
            ) from error
        if not 200 <= response.status_code < 300:
            raise ApiError(method, response.status_code)
        if not response.content:
            return {}
        try:
            result = response.json()
        except ValueError:
            raise RuntimeError(f"{method} returned an invalid JSON response.") from None
        if not isinstance(result, dict):
            raise RuntimeError(f"{method} returned an unexpected JSON response shape.")
        return result


def print_result(label: str, value: dict[str, Any]) -> None:
    """Print an API response without escaping IPA symbols."""
    print(f"\n{label}")
    print(json.dumps(value, indent=2, ensure_ascii=False))


def require_etag(dictionary: dict[str, Any]) -> None:
    if not isinstance(dictionary.get("etag"), str) or not dictionary["etag"]:
        raise ValueError("A current, non-empty dictionary ETag is required.")


def dictionary_state(dictionary: dict[str, Any]) -> tuple[Any, Any]:
    entries = dictionary.get("pronunciations")
    if isinstance(entries, list) and all(isinstance(entry, dict) for entry in entries):
        entries = [
            {
                "displayHeadword": entry.get("displayHeadword"),
                "languageCode": entry.get("languageCode"),
                "phoneSymbols": entry.get("phoneSymbols"),
            }
            for entry in entries
        ]
    return dictionary.get("displayName"), entries


def run_lifecycle(client: PronunciationDictionariesClient) -> None:
    """Run create, list, get, update, and delete with cleanup on failure."""
    dictionary: dict[str, Any] | None = None
    created_name: str | None = None
    deleted = False
    run_id = uuid4().hex
    initial_name = f"API example {run_id}"
    updated_name = f"API example updated {run_id}"
    owned_states = [(initial_name, INITIAL_PRONUNCIATIONS)]
    print(f"Creating dictionary with displayName: {initial_name}")

    try:
        dictionary = client.create_dictionary(
            display_name=initial_name,
            pronunciations=INITIAL_PRONUNCIATIONS,
        )
        client._resource_url(dictionary["name"])
        created_name = dictionary["name"]
        print_result("1. Created dictionary", dictionary)

        page = client.list_dictionaries(page_size=5)
        print_result("2. Listed dictionaries (first page)", page)

        dictionary = client.get_dictionary(dictionary["name"])
        if (
            dictionary.get("name") != created_name
            or dictionary_state(dictionary) != owned_states[0]
        ):
            raise RuntimeError(
                "Dictionary changed unexpectedly; refusing to update it."
            )
        print_result("3. Retrieved dictionary", dictionary)

        owned_states.append((updated_name, REPLACEMENT_PRONUNCIATIONS))
        dictionary = client.update_dictionary(
            dictionary=dictionary,
            display_name=updated_name,
            pronunciations=REPLACEMENT_PRONUNCIATIONS,
        )
        if (
            dictionary.get("name") != created_name
            or dictionary_state(dictionary) != owned_states[1]
        ):
            raise RuntimeError(
                "Updated dictionary does not match the intended contents."
            )
        print_result("4. Replaced dictionary contents", dictionary)

        delete_response = client.delete_dictionary(dictionary)
        print_result("5. Deleted dictionary", delete_response)
        deleted = True
    finally:
        if created_name is not None and not deleted:
            try:
                current = client.get_dictionary(created_name)
                if (
                    current.get("name") != created_name
                    or dictionary_state(current) not in owned_states
                ):
                    raise RuntimeError(
                        "Dictionary changed outside this run; refusing deletion."
                    )
                client.delete_dictionary(current)
                print("\nCleaned up the dictionary after an incomplete run.")
            except ApiError as cleanup_error:
                if cleanup_error.status_code != 404:
                    print(
                        f"\nCleanup failed; inspect {created_name} manually: {cleanup_error}",
                        file=sys.stderr,
                    )
            except (
                requests.RequestException,
                RuntimeError,
                ValueError,
                KeyError,
            ) as cleanup_error:
                print(
                    f"\nCleanup failed; inspect {created_name} manually: "
                    f"{cleanup_error}",
                    file=sys.stderr,
                )
        elif created_name is None:
            # An unsuccessful create may have committed without returning its name.
            if sys.exc_info()[0] is not None:
                print(
                    f"Inspect the dictionary list for displayName {initial_name!r} "
                    "before rerunning; create may have committed.",
                    file=sys.stderr,
                )


def require_environment_variable(name: str) -> str:
    """Return a required environment variable or raise a focused error."""
    value = os.getenv(name)
    if not value:
        raise ValueError(
            f"{name} is not set. Copy .env.example to .env and provide a value."
        )
    return value


def client_from_environment() -> PronunciationDictionariesClient:
    """Load script-local configuration only when an entrypoint is executed."""
    try:
        from dotenv import load_dotenv

        load_dotenv(Path(__file__).with_name(".env"))
    except ImportError:
        pass
    return PronunciationDictionariesClient(
        api_key=require_environment_variable("INWORLD_API_KEY"),
        workspace_id=require_environment_variable("INWORLD_WORKSPACE_ID"),
        api_base_url=os.getenv("INWORLD_API_BASE_URL", DEFAULT_API_BASE_URL),
    )


def main() -> int:
    """Load configuration and run the example."""
    try:
        with client_from_environment() as client:
            run_lifecycle(client)
    except (requests.RequestException, RuntimeError, ValueError, KeyError) as error:
        print(f"Pronunciation dictionary example failed: {error}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tts-pronunciations/js/example_pronunciation_dictionaries.js
#!/usr/bin/env node
/** Manage a named pronunciation dictionary through Inworld's public REST API. */

const { randomUUID } = require('node:crypto');
const { isDeepStrictEqual } = require('node:util');
const path = require('node:path');

const DEFAULT_API_BASE_URL = 'https://api.inworld.ai';
const REQUEST_TIMEOUT_MILLISECONDS = 30_000;
const INITIAL_PRONUNCIATIONS = [
    { displayHeadword: 'Cat', languageCode: 'en-US', phoneSymbols: ['k', 'æ', 't'] },
    { displayHeadword: 'Ship', languageCode: 'en-US', phoneSymbols: ['ʃ', 'ɪ', 'p'] },
];
const REPLACEMENT_PRONUNCIATIONS = [
    { displayHeadword: 'Red', languageCode: 'en-US', phoneSymbols: ['ɹ', 'ɛ', 'd'] },
];

class ApiError extends Error {
    constructor(method, status) {
        super(`${method} request failed with HTTP ${status}.`);
        this.status = status;
    }
}

class PronunciationDictionariesClient {
    constructor(apiKey, workspaceId, apiBaseUrl, { fetchImpl = fetch } = {}) {
        const base = new URL(apiBaseUrl);
        if (base.username || base.password || base.search || base.hash ||
            base.pathname !== '/' || (base.protocol !== 'https:' &&
                !(base.protocol === 'http:' && ['localhost', '127.0.0.1', '[::1]'].includes(base.hostname)))) {
            throw new Error('Use an HTTPS regional API origin (HTTP is allowed only for localhost).');
        }
        this.apiKey = apiKey;
        this.workspaceId = workspaceId;
        this.apiBaseUrl = base.origin;
        this.fetchImpl = fetchImpl;
        this.collectionPath =
            `/pronunciations/v1/workspaces/${encodeURIComponent(workspaceId)}` +
            '/pronunciationDictionaries';
    }

    createDictionary(displayName, pronunciations) {
        return this.request(this.collectionPath, {
            method: 'POST', body: { displayName, pronunciations },
        });
    }

    listDictionaries(pageSize = 5) {
        return this.request(this.collectionPath, { query: { pageSize: String(pageSize) } });
    }

    getDictionary(name) {
        return this.request(this.resourcePath(name));
    }

    updateDictionary(dictionary, displayName, pronunciations) {
        requireEtag(dictionary);
        return this.request(this.resourcePath(dictionary.name), {
            method: 'PATCH',
            query: { updateMask: 'displayName,pronunciations' },
            body: { displayName, pronunciations, etag: dictionary.etag },
        });
    }

    deleteDictionary(dictionary) {
        requireEtag(dictionary);
        return this.request(this.resourcePath(dictionary.name), {
            method: 'DELETE', query: { etag: dictionary.etag },
        });
    }

    synthesize(body) {
        return this.request('/tts/v1/voice', { method: 'POST', body });
    }

    resourcePath(name) {
        const prefix = `workspaces/${this.workspaceId}/pronunciationDictionaries/`;
        const id = typeof name === 'string' && name.startsWith(prefix) ? name.slice(prefix.length) : '';
        if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(id)) {
            throw new Error('Use the full dictionary name returned by create, in INWORLD_WORKSPACE_ID.');
        }
        return `/pronunciations/v1/${name.split('/').map(encodeURIComponent).join('/')}`;
    }

    async request(endpoint, { method = 'GET', query = {}, body } = {}) {
        const url = new URL(`${this.apiBaseUrl}${endpoint}`);
        for (const [name, value] of Object.entries(query)) url.searchParams.set(name, value);
        let response;
        let responseText;
        try {
            response = await this.fetchImpl(url, {
                method,
                redirect: 'error',
                signal: AbortSignal.timeout(REQUEST_TIMEOUT_MILLISECONDS),
                headers: {
                    Authorization: `Basic ${this.apiKey}`,
                    ...(body ? { 'Content-Type': 'application/json' } : {}),
                },
                ...(body ? { body: JSON.stringify(body) } : {}),
            });
            if (!response.ok) throw new ApiError(method, response.status);
            responseText = await response.text();
        } catch (error) {
            if (error instanceof ApiError) throw error;
            throw new Error(`${method} request failed or timed out; its outcome may be unknown. Do not retry a mutation blindly.`);
        }
        try {
            return responseText ? JSON.parse(responseText) : {};
        } catch (_) {
            throw new Error(`${method} returned invalid JSON; its outcome may be unknown.`);
        }
    }
}

function requireEtag(dictionary) {
    if (typeof dictionary.etag !== 'string' || !dictionary.etag) {
        throw new Error('A current, non-empty dictionary ETag is required.');
    }
}

function dictionaryState(dictionary) {
    return {
        displayName: dictionary.displayName,
        pronunciations: dictionary.pronunciations?.map(({ displayHeadword, languageCode, phoneSymbols }) =>
            ({ displayHeadword, languageCode, phoneSymbols })),
    };
}

async function cleanupOwnedDictionary(client, name, expectedStates) {
    let current;
    try {
        current = await client.getDictionary(name);
    } catch (error) {
        if (error instanceof ApiError && error.status === 404) return;
        throw error;
    }
    // A lost PATCH response may leave either the preimage or intended postimage.
    // Refuse to delete content changed by another writer; the fresh ETag fences races.
    if (current.name !== name || !expectedStates.some((state) =>
        isDeepStrictEqual(dictionaryState(current), state))) {
        throw new Error('Cleanup refused: dictionary contents changed unexpectedly. Inspect it manually.');
    }
    await client.deleteDictionary(current);
}

function requireEnvironmentVariable(name) {
    const value = process.env[name];
    if (!value) throw new Error(`${name} is not set. Copy .env.example to .env and provide a value.`);
    return value;
}

function clientFromEnvironment() {
    // Importing this module must not load credentials or execute requests.
    require('dotenv').config({ path: path.join(__dirname, '.env') });
    return new PronunciationDictionariesClient(
        requireEnvironmentVariable('INWORLD_API_KEY'),
        requireEnvironmentVariable('INWORLD_WORKSPACE_ID'),
        process.env.INWORLD_API_BASE_URL || DEFAULT_API_BASE_URL,
    );
}

async function runLifecycle(client, { log = console.log, error = console.error, id = randomUUID() } = {}) {
    const initial = { displayName: `API example ${id}`, pronunciations: INITIAL_PRONUNCIATIONS };
    const replacement = { displayName: `API example updated ${id}`, pronunciations: REPLACEMENT_PRONUNCIATIONS };
    let name;
    const expectedStates = [dictionaryState(initial)];
    log(`Creating disposable dictionary: ${initial.displayName}`);
    try {
        let dictionary = await client.createDictionary(initial.displayName, initial.pronunciations);
        client.resourcePath(dictionary.name);
        name = dictionary.name;
        log('1. Created dictionary', dictionary);
        log('2. Listed dictionaries (first page)', await client.listDictionaries(5));
        dictionary = await client.getDictionary(name);
        if (dictionary.name !== name || !isDeepStrictEqual(dictionaryState(dictionary), expectedStates[0])) {
            throw new Error('Dictionary changed unexpectedly; refusing to update it.');
        }
        log('3. Retrieved dictionary', dictionary);
        expectedStates.push(dictionaryState(replacement));
        dictionary = await client.updateDictionary(dictionary, replacement.displayName, replacement.pronunciations);
        if (dictionary.name !== name || !isDeepStrictEqual(dictionaryState(dictionary), expectedStates[1])) {
            throw new Error('Updated dictionary does not match the intended contents.');
        }
        log('4. Replaced dictionary contents', dictionary);
        await client.deleteDictionary(dictionary);
        name = null;
        log('5. Deleted dictionary');
    } catch (failure) {
        if (!name) error(`If create committed without a usable response, find displayName "${initial.displayName}" and reconcile manually.`);
        throw failure;
    } finally {
        if (name) {
            try {
                await cleanupOwnedDictionary(client, name, expectedStates);
                log('Cleaned up the dictionary after an incomplete run.');
            } catch (cleanupError) {
                error(`Cleanup failed for ${name}; inspect and delete it manually: ${cleanupError.message}`);
            }
        }
    }
}

if (require.main === module) {
    Promise.resolve().then(() => runLifecycle(clientFromEnvironment())).catch((error) => {
        console.error(`Pronunciation dictionary example failed: ${error.message}`);
        process.exitCode = 1;
    });
}

module.exports = {
    ApiError, PronunciationDictionariesClient, clientFromEnvironment,
    requireEnvironmentVariable, cleanupOwnedDictionary, runLifecycle,
    INITIAL_PRONUNCIATIONS, REPLACEMENT_PRONUNCIATIONS,
};
